### ISSUE TITLE
Discussion: Content type for watch serialization in protobuf

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#24501]
Currently, we have two serialization protocols in Kube:

1. JSON encoded runtime.Objects - a single JSON object `{...}` on disk or on the wire - GET/LIST/.../error cases.  
     This is clearly content type `application/json`

2. JSON encoded stream of watch.Events separated by whitespace, returned by watch.  
    This, according to all formal definitions, is not JSON (because it's > 1 object) but since the specs do not define a content type it is currently returned as content type `application/json`, and clients have to infer the separation by virtue of calling a "watch" API (GET /resource?watch=1).  

    However, we also have the case where we get an admission / security / auth / format error prior to starting the watch, in which case we return a JSON encoded status object and end the stream (on a non 2xx status code). Clients don't have to work hard to distinguish the object, although they probably should be able to.

HTTP Accept header negotiation is pretty clear cut - a client requests a content-type, and the server must error out or select from their list a matching value.  We *should* unambiguously define the content-type for watch serialization, but if we select a new content-type (application/vnd.kubernetes.json) we will break old clients, which we cannot do.

Options:

1. Introduce a new optional watch JSON type - `application/vnd.kubernetes.watchjson` that well behaved clients can request.  Clients that asked for this type on normal requests would get an error.
2. Modify the returned content type for watch with a mime-qualifier: `application/json;kubernetes=watch` or similar.  Might have backwards compat impacts to JSON.

The problem is compounded in protobuf - the default protobuf serialization will be `application/vnd.kubernetes.protobuf` with the following form:

```
Bytes  0-3    
"k8s" 0x00 <protobuf encoded runtime.Unknown ... <protobuf encoded core object>>
```

However, in order to properly do a watch, we need a streaming serialization (protobuf is not self describing) to identify object boundaries - our format there is a simple 4 byte integer length prefix for each "frame" (like GRPC), and each object will be a protobuf encoded watch.Event.  Every object returned by a watch is an event, so we don't need individual type info

```
4 byte integer defining length
0x00 0x00 0x00 0x00 | <protobuf encoded watch.Event field=object <protobuf encoded runtime.Unknown ... <protobuf encoded core object> >
```

The watch protocol is demonstrably different than the normal serialization.  If the server returns an error, we can't distinguish between the two (since the first frame could be "k8s" 0x00 bytes long).  The standard mechanism is to use the content-type to distinguish the two serializations - and like JSON above we should probably be consistent and define a mechanism.

The advantage of option 2 above is that the client can ask for `application/vnd.kubernetes.protobuf` in the Accept header and get back `application/vnd.kubernetes.protobuf` (Status) or `application/vnd.kubernetes.protobuf;stream=events` (watch stream).  If we used option 1, clients that are making watch calls would have to ask for `application/vnd.kubernetes.protobuf.watch` in a distinct fashion than other cases.

In general, I would expect each "content-type" to have a standard streaming format and a standard non-streaming format.  We do not serialize the streaming format to disk (today), instead we serialize Lists or other types.

**Comment by smarterclayton** [comment]
@wojtek-t regarding streaming serialization and what we have to do with content type.

**Comment by smarterclayton** [comment]
@liggitt

**Comment by wojtek-t** [comment]
Fixed by #24539
